localmassifreader.go

```diff
@@ -0,0 +1,310 @@
+package veracity
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/datatrails/go-datatrails-common/azblob"
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+)
+
+type Opener interface {
+	Open(string) (io.ReadCloser, error)
+}
+
+type DirLister interface {
+	// ListFiles returns list of absolute paths
+	// to files (not subdirectories) in a directory
+	ListFiles(string) ([]string, error)
+}
+
+type LocalMassifReader struct {
+	log              logger.Logger
+	opts             massifs.MassifReaderOptions
+	massifs          map[uint64]string
+	headMassifIndex  uint64
+	firstMassifIndex uint64
+	opener           Opener
+	dirLister        DirLister
+}
+
+func (mr *LocalMassifReader) GetMassif(
+	ctx context.Context, tenantIdentity string, massifIndex uint64,
+	opts ...azblob.Option,
+) (massifs.MassifContext, error) {
+
+	var err error
+	mc := massifs.MassifContext{
+		TenantIdentity: tenantIdentity,
+	}
+	if err = mr.readAndPrepareContext(&mc, massifIndex); err != nil {
+		return massifs.MassifContext{}, err
+	}
+	return mc, nil
+}
+
+func (mr *LocalMassifReader) GetHeadMassif(
+	ctx context.Context, tenantIdentity string,
+	opts ...azblob.Option,
+) (massifs.MassifContext, error) {
+
+	var err error
+	mc := massifs.MassifContext{
+		TenantIdentity: tenantIdentity,
+	}
+	if len(mr.massifs) == 0 {
+		return massifs.MassifContext{}, massifs.ErrMassifNotFound
+	}
+	if err = mr.readAndPrepareContext(&mc, mr.headMassifIndex); err != nil {
+		return massifs.MassifContext{}, err
+	}
+
+	return mc, nil
+}
+
+// GetLazyContext is an optimization for remote massif readers
+// and is therefor not implemented for local massif reader
+func (mr *LocalMassifReader) GetLazyContext(
+	ctx context.Context, tenantIdentity string, which massifs.LogicalBlob,
+	opts ...azblob.Option,
+) (massifs.LogBlobContext, uint64, error) {
+
+	return massifs.LogBlobContext{}, 0, fmt.Errorf("not implemented for local storage")
+}
+
+func (mr *LocalMassifReader) GetFirstMassif(
+	ctx context.Context, tenantIdentity string,
+	opts ...azblob.Option,
+) (massifs.MassifContext, error) {
+
+	var err error
+	mc := massifs.MassifContext{
+		TenantIdentity: tenantIdentity,
+	}
+	if len(mr.massifs) == 0 {
+		return massifs.MassifContext{}, massifs.ErrMassifNotFound
+	}
+	if err = mr.readAndPrepareContext(&mc, mr.firstMassifIndex); err != nil {
+		return massifs.MassifContext{}, err
+	}
+
+	return mc, nil
+}
+
+// NewLocalMassifReader creates MassifReader that reads from
+// local files on disc - it mostly ignores tenant identity
+// as we assume all the logs on the disc are for the tenant one is
+// interested in - but it's still valid to pass tenant ID here
+func NewLocalMassifReader(
+	log logger.Logger, opener Opener, dirLister DirLister, logdir string, logfile string, opts ...massifs.MassifReaderOption,
```

Comment:
is it worth making logdir and logfile options? or having a New() function for each?

There maybe a time when you want one or the other, not both?

Reply:
I just changed the interface here a bit - so now it's only local data and a bool for dir/not dir

Reply:
added options now for directory